src/locale/keymap-util.c

```diff
@@ -780,3 +782,141 @@ int x11_convert_to_vconsole(Context *c) {
+         * a nice config file without empty lines */
+        first_line = true;
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                char *line_locale;
+
+                r = read_line(fr, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                if (locale_enabled) {
+                        /* Just complete writing the file if the new locale was already enabled */
+                        fprintf(fw, "%s%s", first_line? "" : "\n", line);
+                        first_line = false;
+                        continue;
+                }
+
+                line = strstrip(line);
+                if (isempty(line)) {
+                        fputs("\n", fw);
+                        first_line = false;
+                        continue;
+                }
+
+                line_locale = line;
+                if (line_locale[0] == '#')
+                        line_locale = strstrip(line_locale + 1);
+
+                if (strcasecmp_ptr(line_locale, locale_entry) == 0) {
+                        /* Uncomment existing line for new locale */
+                        fprintf(fw, "%s%s", first_line? "" : "\n", locale_entry);
+                        locale_enabled = true;
+                        first_line = false;
+                        continue;
+                }
+
+                /* The line was not for the locale we want to enable, just copy it */
+                fprintf(fw, "%s%s", first_line? "" : "\n", line);
+                first_line = false;
+        }
+
+        /* Add locale to enable to the end of the file if it was not found as commented line */
+        if (!locale_enabled)
+                fprintf(fw, "\n%s", locale_entry);
+        fputs("\n", fw);
+
+        r = fflush_sync_and_check(fw);
+        if (r < 0)
+                goto fail;
+
+        if (rename(temp_path, "/etc/locale.gen") < 0) {
+                r = -errno;
+                goto fail;
+        }
+        temp_path = mfree(temp_path);
+
+        return 0;
+
+fail:
+        if (temp_path)
+                (void) unlink(temp_path);
+        return r;
+}
+
+int locale_gen_run(void) {
+        pid_t pid;
+        int r;
+
+        /* Silently do nothing if locale-gen was not available */
+        if (!locale_gen_check_available())
+                return 0;
+
+        r = safe_fork("(sd-localegen)", FORK_RESET_SIGNALS|FORK_RLIMIT_NOFILE_SAFE|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_LOG, &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                execl("/usr/sbin/locale-gen", "/usr/sbin/locale-gen", NULL);
+                log_error_errno(errno, "Failed to invoke locale-gen: %m");
+                _exit(EXIT_FAILURE);
+        }
+
+        return wait_for_terminate_and_check("/usr/sbin/locale-gen", pid, WAIT_LOG_ABNORMAL);
```

Comment:
instead of this you could also just add FORK_WAIT to the safe_fork() expression above, it will then do the waiting implcitly.

Reply:
But then there is no easier way to get the process' exit code, just in case locale-gen fails, or is there? All other instances of `FORK_WAIT` in systemd's code don't seem to care about the forked processes exist code and just log failure implicitly.

Reply:
do you need the precise error code? i don#t think so?

if you pass FORK_WAIT to safe_fork() then we'll wait for the process to exit, and only do a very basic check for the exit code: if it's zero we'll succeed, and if any other value we'll return this as -EPROTO. Hence, FORK_WAIT is good enough if all you want to check is exit code being zero/non-zero. if you need to have the precise value you cannot use it. but afaics you don't need it here either